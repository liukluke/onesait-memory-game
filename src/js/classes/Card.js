import logoOnesait from '../../assets/image/onesait-logo.svg'

export default class Card {
    constructor(cardImg = '', game = {}, logo = logoOnesait) {
        this.imgFront = cardImg
        this.game = game
        this.imgBack = logo
    }

    createCard() {
        const liCard = document.createElement('li')
        liCard.classList.add('cards')

        const frontCard = document.createElement('img')
        frontCard.classList.add('cards__front')
        frontCard.src = this.imgFront

        const backCard = document.createElement('img')
        backCard.classList.add('cards__back')
        backCard.src = this.imgBack

        new Array(frontCard, backCard).forEach((e) => liCard.appendChild(e))

        liCard.onclick = (e) => this.game.cardSelected(e.target.parentElement)

        return liCard
    }
}
