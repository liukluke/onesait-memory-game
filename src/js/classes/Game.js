import Board from './Board'
import Info from './Info'

export default class Game {
    constructor() {
        this.info = new Info()
        this.board = new Board()
        this.firstCard = undefined
    }
    
    checkCard(secondCard) {
        this.activeGame = false
        this.firstCard.firstElementChild.src === secondCard.firstElementChild.src ? this.matchMethod() : this.noMatchMethod(secondCard)
    }

    matchMethod () {
        this.firstCard = undefined
        this.counter += 1
        if (this.counter >= 6) {
            setTimeout(
                () =>
                    this.startGame([
                        'Has ganado.',
                        '¡Puedes seguir trabajando despues del covid!'
                    ]),
                1000
            )
        } else this.activeGame = true
    }

    noMatchMethod(secondCard) {
        setTimeout(() => {
            new Array(this.firstCard, secondCard).forEach((card) =>
                card.classList.remove('flip')
            )
            this.firstCard = undefined
            this.attemps -= 1
            this.info.updateScores('heart', this.attemps)
            this.attemps < 0 ? this.startGame(['Has perdido.', '¡Te van a hacer un ERTE!']) : this.activeGame = true
        }, 1000)
    }

    cardSelected(card) {
        if (this.activeGame) {
            card.classList.add('flip')
            this.firstCard === undefined
                ? (this.firstCard = card)
                : this.checkCard(card)
        }
    }

    createBoard() {
        this.board.create(this)
        this.info.createScores()
        this.counter = 0
        this.attemps = 3
    }

    startGame(text) {
        const startButton = this.info.message(text)
        startButton.onclick = () => this.createBoard()
        this.activeGame = true
    }
}
