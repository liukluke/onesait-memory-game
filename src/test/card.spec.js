import Card from './../js/classes/Card'
import logoOnesait from '../../assets/image/onesait-logo.svg'
import Game from './../js/classes/Game'

describe('define instance', () => {
    test('Card should be defined', () => {
        expect(Card).toBeDefined()
    })

    test('Card should be a function', () => {
        expect(typeof Card).toBe('function')
    })

    test('new Card should be an instance of Card', () => {
        expect(new Card()).toBeInstanceOf(Card)
    })
})

test('card structure', () => {
    const card = new Card()
    expect(card).toEqual(
        expect.objectContaining({
            imgFront: expect.any(String),
            game: expect.any(Object),
            imgBack: expect.any(Object),
            createCard: expect.any(Function)
        })
    )

    expect(card.imgBack).toEqual(logoOnesait)
})

describe('createCard', () => {
    let card, game, cardElement
    beforeEach(() => {
        game = new Game()
        card = new Card('imgFront', game, logoOnesait)
        cardElement = card.createCard()
    })

    test('Card generates DOM element and matches', () => {
        expect(cardElement).toMatchSnapshot()
    })

    test('spyOn card createCard', () => {
        const spyOnGenerateCard = jest.spyOn(card, 'createCard')
        card.createCard()
        expect(spyOnGenerateCard).toHaveBeenCalled()
    })
    
})
