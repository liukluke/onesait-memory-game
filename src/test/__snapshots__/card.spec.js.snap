// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createCard Card generates DOM element and matches 1`] = `
<li
  class="cards"
>
  <img
    class="cards__front"
    src="imgFront"
  />
  <img
    class="cards__back"
    src="[object Object]"
  />
</li>
`;
